# Advanced Machine Learning & MLOps — Tourism Package Prediction
### "Visit with Us": predicting who will buy the Wellness Tourism Package

**Author:** Prabhudatta Choudhury

| Item | Link |
|---|---|
| GitHub repository | `https://github.com/<your-username>/tourism-mlops` *(replace)* |
| Streamlit Community Cloud app | `https://<your-app>.streamlit.app` *(replace)* |

## 1. Business Context & Objective

"Visit with Us" is launching a **Wellness Tourism Package**. Today, customers to call are picked manually, which is
inconsistent, slow and error-prone, so marketing spend goes to people unlikely to buy.

**Objective:** build and operate an **MLOps pipeline on GitHub** that, on every change, automatically

1. registers and validates the customer dataset,
2. cleans it and creates train / test splits (passed between jobs as workflow artifacts),
3. tunes an **XGBoost** classifier with experiment tracking (**MLflow**) and commits the best model back to `main`,
4. serves the committed model through a **Streamlit** app deployed on **Streamlit Community Cloud**.

The model predicts `ProdTaken` (1 = customer buys the package) **before** the customer is contacted.

### Pipeline architecture

```
 push to main ──► GitHub Actions: pipeline.yml
                    │
   job 1  register-dataset   tourism_project/data_register.py   (validate columns, print summary)
                    │
   job 2  data-prep          tourism_project/prep.py            ──► artifact: train-test-splits
                    │
   job 3  model-training     tourism_project/train.py  ◄── artifact
                    │          GridSearchCV + MLflow logging      ──► artifact: mlflow-tracking
                    │          git commit model/best_model.joblib ──► push to main
                    ▼
   Streamlit Community Cloud  tourism_project/deployment/app.py (auto-redeploys on new commit)
```

## 2. Environment Setup
The notebook builds the full repository on disk with `%%writefile` cells and then executes each pipeline step locally,
exactly as GitHub Actions does, so every output is visible here.

In [1]:
# Libraries used across the pipeline (already pinned in the requirements files written below)
!pip install -q pandas==3.0.5 numpy==2.5.3 scikit-learn==1.9.1 xgboost==3.4.1 joblib==1.6.0 mlflow==3.16.1 streamlit==1.65.0

In [2]:
import os, sys, platform
os.environ["MLFLOW_DISABLE_AGENT_HINT"] = "1"   # silence MLflow console hints
import pandas as pd, numpy as np, sklearn, xgboost, mlflow, streamlit

# Master folder of the project; the GitHub repo root is the current working directory
os.makedirs("tourism_project/data", exist_ok=True)
os.makedirs("tourism_project/model", exist_ok=True)
os.makedirs("tourism_project/deployment", exist_ok=True)
os.makedirs(".github/workflows", exist_ok=True)

print("Python      :", platform.python_version())
print("pandas      :", pd.__version__)
print("scikit-learn:", sklearn.__version__)
print("xgboost     :", xgboost.__version__)
print("mlflow      :", mlflow.__version__)
print("streamlit   :", streamlit.__version__)

Python      : 3.13.16
pandas      : 3.0.5
scikit-learn: 1.9.1
xgboost     : 3.4.1
mlflow      : 3.16.1
streamlit   : 1.65.0


## 3. Data Registration *(rubric: 3 pts)*
A master folder `tourism_project/` with a sub-folder `data/` holds the raw dataset (`tourism.csv`).
`data_register.py` **registers** the dataset: it checks that every expected column is present, that the target is
binary, and prints a summary. If validation fails it exits with a non-zero code, which **stops the GitHub workflow**
before any downstream job runs.

In [3]:
import shutil
# Copy the provided CSV into the repository data folder (in Colab: upload tourism.csv first)
SRC = "tourism.csv"
shutil.copy(SRC, "tourism_project/data/tourism.csv")
print(os.listdir("tourism_project/data"))

['tourism.csv']


In [4]:
%%writefile tourism_project/data_register.py
"""Data registration: validate the raw dataset in tourism_project/data and print a summary."""
import sys
from pathlib import Path
import pandas as pd

DATA_PATH = Path(__file__).parent / "data" / "tourism.csv"

EXPECTED_COLUMNS = [
    "CustomerID", "ProdTaken", "Age", "TypeofContact", "CityTier", "DurationOfPitch",
    "Occupation", "Gender", "NumberOfPersonVisiting", "NumberOfFollowups", "ProductPitched",
    "PreferredPropertyStar", "MaritalStatus", "NumberOfTrips", "Passport",
    "PitchSatisfactionScore", "OwnCar", "NumberOfChildrenVisiting", "Designation", "MonthlyIncome",
]


def main() -> None:
    if not DATA_PATH.exists():
        sys.exit(f"ERROR: dataset not found at {DATA_PATH}")

    df = pd.read_csv(DATA_PATH)

    # 1) schema check: every expected column must be present
    missing = [c for c in EXPECTED_COLUMNS if c not in df.columns]
    if missing:
        sys.exit(f"ERROR: missing expected columns: {missing}")
    extra = [c for c in df.columns if c not in EXPECTED_COLUMNS]

    # 2) target check: ProdTaken must be binary 0/1
    if not set(df["ProdTaken"].dropna().unique()) <= {0, 1}:
        sys.exit("ERROR: target ProdTaken is not binary")

    # 3) summary
    print("=" * 60)
    print("DATASET REGISTERED:", DATA_PATH.name)
    print("=" * 60)
    print(f"Rows x Columns     : {df.shape[0]} x {df.shape[1]}")
    print(f"Expected columns   : all {len(EXPECTED_COLUMNS)} present")
    print(f"Extra columns      : {extra or 'none'}")
    print(f"Duplicate rows     : {df.duplicated().sum()}")
    print(f"Duplicate customers: {df['CustomerID'].duplicated().sum()}")
    print(f"Missing values     : {int(df.isna().sum().sum())}")
    print(f"Target balance     : {df['ProdTaken'].value_counts(normalize=True).round(3).to_dict()}")
    print("\nColumn types:")
    print(df[EXPECTED_COLUMNS].dtypes.to_string())
    print("\nNumeric summary:")
    print(df[EXPECTED_COLUMNS].describe().T.round(2).to_string())
    print("\nValidation PASSED")


if __name__ == "__main__":
    main()

Writing tourism_project/data_register.py


In [5]:
!python tourism_project/data_register.py

DATASET REGISTERED: tourism.csv
Rows x Columns     : 4128 x 21
Expected columns   : all 20 present
Extra columns      : ['Unnamed: 0']
Duplicate rows     : 0
Duplicate customers: 0
Missing values     : 0
Target balance     : {0: 0.807, 1: 0.193}

Column types:
CustomerID                    int64
ProdTaken                     int64
Age                         float64
TypeofContact                   str
CityTier                      int64
DurationOfPitch             float64
Occupation                      str
Gender                          str
NumberOfPersonVisiting        int64
NumberOfFollowups           float64
ProductPitched                  str
PreferredPropertyStar       float64
MaritalStatus                   str
NumberOfTrips               float64
Passport                      int64
PitchSatisfactionScore        int64
OwnCar                        int64
NumberOfChildrenVisiting    float64
Designation                     str
MonthlyIncome               float64

Numeric summary:
 

**Observations**
* All 20 expected columns are present; the only extra column is an unnamed index column left over from a previous export — it carries no information and is dropped in data preparation.
* There are no missing values and no duplicate customers.
* The target is **imbalanced (~19 % buyers)**, so plain accuracy would be misleading: the model uses `scale_pos_weight` and is tuned on **F1**, and recall/precision on the buyer class are reported.

In [6]:
# Quick look at categorical levels to decide on cleaning rules
raw = pd.read_csv("tourism_project/data/tourism.csv")
for c in raw.select_dtypes(exclude="number").columns:
    print(f"{c:15s}", raw[c].value_counts().to_dict())

TypeofContact   {'Self Enquiry': 2918, 'Company Invited': 1210}
Occupation      {'Salaried': 1999, 'Small Business': 1746, 'Large Business': 381, 'Free Lancer': 2}
Gender          {'Male': 2463, 'Female': 1510, 'Fe Male': 155}
ProductPitched  {'Basic': 1615, 'Deluxe': 1422, 'Standard': 737, 'Super Deluxe': 250, 'King': 104}
MaritalStatus   {'Married': 1990, 'Divorced': 789, 'Unmarried': 682, 'Single': 667}
Designation     {'Executive': 1615, 'Manager': 1422, 'Senior Manager': 737, 'AVP': 250, 'VP': 104}


**Observations**
* `Gender` contains a typo level **"Fe Male"** (155 rows) → merged into `Female`.
* `Occupation = "Free Lancer"` has only 2 rows; it is kept, and the encoder is configured to ignore unseen levels so the model is robust either way.
* `ProductPitched` and `Designation` map one-to-one (Basic↔Executive, Deluxe↔Manager…), i.e. the pitched product was chosen by designation. Both are kept: tree models are unaffected by this redundancy.

## 4. Data Preparation *(rubric: 7 pts)*
`prep.py`:
1. loads the dataset **directly from the repository data folder** (`tourism_project/data/tourism.csv`),
2. cleans it — drops the unnamed index and `CustomerID` (identifiers, no predictive value), fixes the `Fe Male` typo, removes duplicates, and imputes any missing values (median / mode) so future data drops don't break the pipeline,
3. performs a stratified 80 / 20 train-test split,
4. saves `Xtrain.csv`, `Xtest.csv`, `ytrain.csv`, `ytest.csv` locally to `tourism_project/data/processed/`.

In GitHub Actions that folder is uploaded with `actions/upload-artifact` as **`train-test-splits`**, which the training job downloads.

In [7]:
%%writefile tourism_project/prep.py
"""Data preparation: clean the raw dataset and create train/test splits."""
from pathlib import Path
import pandas as pd
from sklearn.model_selection import train_test_split

BASE = Path(__file__).parent
DATA_PATH = BASE / "data" / "tourism.csv"         # loaded directly from the repo data folder
OUT_DIR = BASE / "data" / "processed"              # uploaded as a workflow artifact
TARGET = "ProdTaken"


def clean(df: pd.DataFrame) -> pd.DataFrame:
    # Drop identifier / index columns: they carry no predictive signal
    df = df.drop(columns=[c for c in df.columns if c.startswith("Unnamed")] + ["CustomerID"])

    # Fix inconsistent category spelling
    df["Gender"] = df["Gender"].replace({"Fe Male": "Female"})

    # Remove exact duplicate records
    df = df.drop_duplicates()

    # Impute any missing values (defensive: the current file has none)
    for col in df.columns:
        if df[col].isna().any():
            fill = df[col].median() if pd.api.types.is_numeric_dtype(df[col]) else df[col].mode()[0]
            df[col] = df[col].fillna(fill)
    return df


def main() -> None:
    df = pd.read_csv(DATA_PATH)
    print("Raw shape    :", df.shape)
    df = clean(df)
    print("Cleaned shape:", df.shape)

    X, y = df.drop(columns=[TARGET]), df[TARGET]
    Xtrain, Xtest, ytrain, ytest = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )

    OUT_DIR.mkdir(parents=True, exist_ok=True)
    Xtrain.to_csv(OUT_DIR / "Xtrain.csv", index=False)
    Xtest.to_csv(OUT_DIR / "Xtest.csv", index=False)
    ytrain.to_csv(OUT_DIR / "ytrain.csv", index=False)
    ytest.to_csv(OUT_DIR / "ytest.csv", index=False)

    print(f"Train: {Xtrain.shape}  buyers={ytrain.mean():.3f}")
    print(f"Test : {Xtest.shape}  buyers={ytest.mean():.3f}")
    print("Saved splits to", OUT_DIR, "->", sorted(p.name for p in OUT_DIR.iterdir()))


if __name__ == "__main__":
    main()

Writing tourism_project/prep.py


In [8]:
!python tourism_project/prep.py

Raw shape    : (4128, 21)
Cleaned shape: (4011, 19)


Train: (3208, 18)  buyers=0.193
Test : (803, 18)  buyers=0.193
Saved splits to /home/claude/repo/tourism_project/data/processed -> ['Xtest.csv', 'Xtrain.csv', 'ytest.csv', 'ytrain.csv']


**Observations**
* 2 columns were dropped (index + `CustomerID`), leaving 18 features + target.
* **117 duplicate records** surfaced once `CustomerID` was removed (identical customers under different IDs). Dropping them *before* the split prevents the same record landing in both train and test, which would inflate test scores.
* Stratification keeps the buyer rate identical (~19.3 %) in train and test, so test metrics are representative.

## 5. Model Building with Experimentation Tracking *(rubric: 13 pts)*
`train.py`:
* **loads train/test data from the workflow artifact** (`ARTIFACT_DIR`, defaulting to the folder the artifact is downloaded into),
* defines a scikit-learn `Pipeline`: `OneHotEncoder` for categoricals + numeric passthrough → **XGBoost** classifier with `scale_pos_weight` to counter class imbalance,
* defines a parameter grid and **tunes** it with 5-fold stratified `GridSearchCV` (scoring = F1),
* **logs every tuned parameter combination** to MLflow as a nested run (params + mean/std CV F1), and logs the best params, test metrics and the model in the parent run,
* evaluates the best model on train and test,
* saves the best model to `tourism_project/model/best_model.joblib`, which the workflow **commits to the repository**.

The whole preprocessing is inside the saved pipeline, so the Streamlit app sends raw inputs and gets consistent predictions.

In [9]:
%%writefile tourism_project/train.py
"""Model training: tune XGBoost, track experiments with MLflow, save the best model."""
import json
import os
from pathlib import Path

import joblib
import mlflow
import pandas as pd
from sklearn.compose import ColumnTransformer
from sklearn.metrics import (accuracy_score, classification_report, f1_score,
                             precision_score, recall_score, roc_auc_score)
from sklearn.model_selection import GridSearchCV, StratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder
from xgboost import XGBClassifier

BASE = Path(__file__).parent
ARTIFACT_DIR = Path(os.getenv("ARTIFACT_DIR", BASE / "data" / "processed"))
MODEL_DIR = BASE / "model"

mlflow.set_tracking_uri(os.getenv("MLFLOW_TRACKING_URI", "sqlite:///mlflow.db"))
mlflow.set_experiment("tourism-package-prediction")

# ---------- 1. Load train/test splits from the workflow artifact
Xtrain = pd.read_csv(ARTIFACT_DIR / "Xtrain.csv")
Xtest = pd.read_csv(ARTIFACT_DIR / "Xtest.csv")
ytrain = pd.read_csv(ARTIFACT_DIR / "ytrain.csv").squeeze("columns")
ytest = pd.read_csv(ARTIFACT_DIR / "ytest.csv").squeeze("columns")
print(f"Loaded splits from {ARTIFACT_DIR}: train={Xtrain.shape}, test={Xtest.shape}")

# ---------- 2. Define model
cat_cols = Xtrain.select_dtypes(exclude="number").columns.tolist()
num_cols = Xtrain.select_dtypes(include="number").columns.tolist()

preprocessor = ColumnTransformer([
    ("cat", OneHotEncoder(handle_unknown="ignore"), cat_cols),
    ("num", "passthrough", num_cols),
])

pos_weight = (ytrain == 0).sum() / (ytrain == 1).sum()   # counter the ~19% positive class
model = Pipeline([
    ("prep", preprocessor),
    ("xgb", XGBClassifier(scale_pos_weight=pos_weight, eval_metric="logloss",
                          random_state=42, n_jobs=-1)),
])

# ---------- 3. Define tuning parameters
param_grid = {
    "xgb__n_estimators": [100, 200],
    "xgb__max_depth": [4, 6],
    "xgb__learning_rate": [0.05, 0.1],
    "xgb__subsample": [0.8, 1.0],
    "xgb__colsample_bytree": [0.7, 1.0],
}

with mlflow.start_run(run_name="xgb-gridsearch"):
    # ---------- 4. Tune
    search = GridSearchCV(model, param_grid, scoring="f1", n_jobs=-1,
                          cv=StratifiedKFold(5, shuffle=True, random_state=42))
    search.fit(Xtrain, ytrain)

    # ---------- 5. Log ALL tuned parameter combinations as nested runs
    res = search.cv_results_
    for i, params in enumerate(res["params"]):
        with mlflow.start_run(run_name=f"candidate-{i:02d}", nested=True):
            mlflow.log_params({k.replace("xgb__", ""): v for k, v in params.items()})
            mlflow.log_metric("cv_f1_mean", res["mean_test_score"][i])
            mlflow.log_metric("cv_f1_std", res["std_test_score"][i])
    print(f"Logged {len(res['params'])} candidate parameter sets to MLflow")

    best = search.best_estimator_
    mlflow.log_params({f"best_{k.replace('xgb__', '')}": v for k, v in search.best_params_.items()})
    mlflow.log_param("scale_pos_weight", round(float(pos_weight), 3))
    mlflow.log_metric("best_cv_f1", search.best_score_)

    # ---------- 6. Evaluate
    metrics = {}
    for split, X, y in [("train", Xtrain, ytrain), ("test", Xtest, ytest)]:
        pred, proba = best.predict(X), best.predict_proba(X)[:, 1]
        metrics.update({
            f"{split}_accuracy": accuracy_score(y, pred),
            f"{split}_precision": precision_score(y, pred),
            f"{split}_recall": recall_score(y, pred),
            f"{split}_f1": f1_score(y, pred),
            f"{split}_roc_auc": roc_auc_score(y, proba),
        })
    mlflow.log_metrics(metrics)

    print("Best params :", search.best_params_)
    print(f"Best CV F1  : {search.best_score_:.4f}")
    print(pd.DataFrame({s: {m: metrics[f'{s}_{m}'] for m in
                            ['accuracy', 'precision', 'recall', 'f1', 'roc_auc']}
                        for s in ['train', 'test']}).round(4).to_string())
    print("\nTest classification report:\n", classification_report(ytest, best.predict(Xtest)))

    # ---------- 7. Save best model (committed to the repo by the workflow)
    MODEL_DIR.mkdir(exist_ok=True)
    joblib.dump(best, MODEL_DIR / "best_model.joblib")
    (MODEL_DIR / "metrics.json").write_text(json.dumps(
        {"best_params": search.best_params_, **{k: round(v, 4) for k, v in metrics.items()}}, indent=2))
    mlflow.log_artifact(str(MODEL_DIR / "best_model.joblib"), artifact_path="model")
    print("Saved best model to", MODEL_DIR / "best_model.joblib")

Writing tourism_project/train.py


In [10]:
# Locally the "artifact" is the folder prep.py wrote; in GitHub Actions it is downloaded to the same path.
!ARTIFACT_DIR=tourism_project/data/processed python tourism_project/train.py

2026/10/05 04:19:24 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/10/05 04:19:24 INFO mlflow.store.db.utils: Updating database tables


2026/10/05 04:19:25 INFO mlflow.tracking.fluent: Experiment with name 'tourism-package-prediction' does not exist. Creating a new experiment.


Loaded splits from tourism_project/data/processed: train=(3208, 18), test=(803, 18)


Logged 32 candidate parameter sets to MLflow


Best params : {'xgb__colsample_bytree': 0.7, 'xgb__learning_rate': 0.1, 'xgb__max_depth': 6, 'xgb__n_estimators': 200, 'xgb__subsample': 0.8}
Best CV F1  : 0.7786
            train    test
accuracy   0.9981  0.9253
precision  0.9904  0.8322
recall     1.0000  0.7677
f1         0.9952  0.7987
roc_auc    1.0000  0.9429

Test classification report:
               precision    recall  f1-score   support

           0       0.95      0.96      0.95       648
           1       0.83      0.77      0.80       155

    accuracy                           0.93       803
   macro avg       0.89      0.87      0.88       803
weighted avg       0.92      0.93      0.92       803



Saved best model to /home/claude/repo/tourism_project/model/best_model.joblib


### 5.1 Experiment tracking — what MLflow recorded

In [11]:
import mlflow
mlflow.set_tracking_uri("sqlite:///mlflow.db")
runs = mlflow.search_runs(experiment_names=["tourism-package-prediction"])
cand = runs[runs["tags.mlflow.runName"].str.startswith("candidate")]
cols = ["tags.mlflow.runName"] + [c for c in cand.columns if c.startswith("params.")] + ["metrics.cv_f1_mean", "metrics.cv_f1_std"]
print(f"Total runs logged: {len(runs)}  (1 parent + {len(cand)} tuned candidates)")
cand[cols].dropna(axis=1, how="all").sort_values("metrics.cv_f1_mean", ascending=False).head(10).round(4)

Total runs logged: 33  (1 parent + 32 tuned candidates)


,tags.mlflow.runName,params.subsample,params.n_estimators,params.max_depth,params.learning_rate,params.colsample_bytree,metrics.cv_f1_mean,metrics.cv_f1_std
17,candidate-14,0.8,200,6,0.1,0.7,0.7786,0.0259
1,candidate-30,0.8,200,6,0.1,1.0,0.7716,0.0198
16,candidate-15,1.0,200,6,0.1,0.7,0.7635,0.0390
8,candidate-23,1.0,200,6,0.05,1.0,0.7563,0.0236
9,candidate-22,0.8,200,6,0.05,1.0,0.7552,0.0246
0,candidate-31,1.0,200,6,0.1,1.0,0.7544,0.0217
19,candidate-12,0.8,100,6,0.1,0.7,0.7509,0.0379
18,candidate-13,1.0,100,6,0.1,0.7,0.7507,0.0287
24,candidate-07,1.0,200,6,0.05,0.7,0.7481,0.0318
25,candidate-06,0.8,200,6,0.05,0.7,0.7456,0.0241


In [12]:
# Feature importance of the best model (aggregated per original feature)
import joblib, matplotlib.pyplot as plt
best = joblib.load("tourism_project/model/best_model.joblib")
names = best.named_steps["prep"].get_feature_names_out()
imp = pd.Series(best.named_steps["xgb"].feature_importances_, index=names)
imp.index = [n.split("__", 1)[1] for n in imp.index]
orig = pd.read_csv("tourism_project/data/processed/Xtrain.csv").columns
agg = pd.Series({f: imp[[i for i in imp.index if i == f or i.startswith(f + "_")]].sum() for f in orig}).sort_values()
agg.plot.barh(figsize=(8, 6), title="XGBoost feature importance (gain share)"); plt.tight_layout(); plt.show()

**Observations**
* All 32 parameter combinations were logged as nested MLflow runs along with the best parameters, the class weight and train/test metrics.
* The best model's test metrics are printed above; see **Section 9** for the business interpretation.
* In CI, `mlflow.db` and `mlruns/` are uploaded as the `mlflow-tracking` artifact so every pipeline run keeps its full experiment history (open it with `mlflow ui --backend-store-uri sqlite:///mlflow.db`).

## 6. Model Deployment *(rubric: 11 pts)*
* `deployment/app.py` — Streamlit app that **loads the model committed to the repository** (`tourism_project/model/best_model.joblib`), collects customer inputs, **saves them into a DataFrame** and returns the purchase probability.
* `deployment/requirements.txt` — dependency file. Streamlit Community Cloud looks for it next to the app entry-point first, and the versions match training exactly so the pickled pipeline loads cleanly.

In [13]:
%%writefile tourism_project/deployment/app.py
"""Streamlit front-end for the Wellness Tourism Package purchase model."""
from pathlib import Path

import joblib
import pandas as pd
import streamlit as st

MODEL_PATH = Path(__file__).resolve().parents[1] / "model" / "best_model.joblib"


@st.cache_resource
def load_model():
    return joblib.load(MODEL_PATH)        # model committed to the repo by the CI pipeline


model = load_model()

st.set_page_config(page_title="Wellness Package Predictor", page_icon="🧳")
st.title("🧳 Wellness Tourism Package — Purchase Predictor")
st.write("Enter the customer details to estimate whether they will buy the **Wellness Tourism Package** "
         "before the sales team contacts them.")

c1, c2 = st.columns(2)
with c1:
    st.subheader("Customer")
    age = st.number_input("Age", 18, 100, 35)
    gender = st.selectbox("Gender", ["Male", "Female"])
    marital = st.selectbox("Marital status", ["Married", "Single", "Unmarried", "Divorced"])
    occupation = st.selectbox("Occupation", ["Salaried", "Small Business", "Large Business", "Free Lancer"])
    designation = st.selectbox("Designation", ["Executive", "Manager", "Senior Manager", "AVP", "VP"])
    income = st.number_input("Monthly income", 1000, 1_000_000, 22000, step=500)
    city_tier = st.selectbox("City tier", [1, 2, 3])
    passport = st.selectbox("Has passport?", ["Yes", "No"])
    own_car = st.selectbox("Owns a car?", ["Yes", "No"])
with c2:
    st.subheader("Trip & interaction")
    persons = st.number_input("Number of persons visiting", 1, 10, 3)
    children = st.number_input("Children (< 5 yrs) visiting", 0, 5, 1)
    trips = st.number_input("Average trips per year", 0, 30, 3)
    star = st.selectbox("Preferred property star", [3, 4, 5])
    contact = st.selectbox("Type of contact", ["Self Enquiry", "Company Invited"])
    product = st.selectbox("Product pitched", ["Basic", "Standard", "Deluxe", "Super Deluxe", "King"])
    duration = st.number_input("Duration of pitch (min)", 1, 150, 15)
    followups = st.number_input("Number of follow-ups", 0, 10, 4)
    pitch_score = st.slider("Pitch satisfaction score", 1, 5, 3)

# Collect inputs into a DataFrame with the exact training schema
input_df = pd.DataFrame([{
    "Age": age, "TypeofContact": contact, "CityTier": city_tier, "DurationOfPitch": duration,
    "Occupation": occupation, "Gender": gender, "NumberOfPersonVisiting": persons,
    "NumberOfFollowups": followups, "ProductPitched": product, "PreferredPropertyStar": star,
    "MaritalStatus": marital, "NumberOfTrips": trips, "Passport": int(passport == "Yes"),
    "PitchSatisfactionScore": pitch_score, "OwnCar": int(own_car == "Yes"),
    "NumberOfChildrenVisiting": children, "Designation": designation, "MonthlyIncome": income,
}])

with st.expander("Input data sent to the model"):
    st.dataframe(input_df)

if st.button("Predict", type="primary"):
    proba = float(model.predict_proba(input_df)[0, 1])
    if proba >= 0.5:
        st.success(f"✅ Likely to BUY the package — probability {proba:.1%}. Prioritise this customer.")
    else:
        st.warning(f"❌ Unlikely to buy — probability {proba:.1%}. Lower priority for outreach.")

Writing tourism_project/deployment/app.py


In [14]:
%%writefile tourism_project/deployment/requirements.txt
pandas==3.0.5
numpy==2.5.3
scikit-learn==1.9.1
xgboost==3.4.1
joblib==1.6.0
streamlit==1.65.0

Writing tourism_project/deployment/requirements.txt


### 6.1 Smoke-test the app headlessly (same code that runs on Streamlit Cloud)

In [15]:
from streamlit.testing.v1 import AppTest
at = AppTest.from_file(os.path.abspath("tourism_project/deployment/app.py"), default_timeout=60).run()
at.button[0].click().run()
print("Exceptions :", list(at.exception) or "none")
print("Prediction :", (at.success or at.warning)[0].value)

2026-10-05 04:19:41.383 WARNING streamlit.runtime.scriptrunner_utils.script_run_context: Thread 'MainThread': missing ScriptRunContext! This warning can be ignored when running in bare mode.


Exceptions : none
Prediction : Unlikely to buy — probability 15.6%. Lower priority for outreach.


In [16]:
# Same prediction straight from the committed model, for two contrasting customers
sample = pd.read_csv("tourism_project/data/processed/Xtest.csv")
ytest = pd.read_csv("tourism_project/data/processed/ytest.csv").squeeze("columns")
demo = sample.iloc[[ytest[ytest == 1].index[0], ytest[ytest == 0].index[0]]]
demo.assign(actual=ytest.loc[demo.index].values, buy_probability=best.predict_proba(demo)[:, 1].round(3))

,Age,TypeofContact,CityTier,DurationOfPitch,Occupation,Gender,NumberOfPersonVisiting,NumberOfFollowups,ProductPitched,PreferredPropertyStar,MaritalStatus,NumberOfTrips,Passport,PitchSatisfactionScore,OwnCar,NumberOfChildrenVisiting,Designation,MonthlyIncome,actual,buy_probability
0,50.0,Company Invited,3,14.0,Large Business,Male,2,3.0,Deluxe,3.0,Divorced,4.0,1,5,1,0.0,Manager,21796.0,1,0.914
1,35.0,Self Enquiry,1,15.0,Small Business,Male,3,2.0,Deluxe,3.0,Married,4.0,0,3,0,2.0,Manager,23082.0,0,0.019


## 7. MLOps Pipeline with GitHub Actions *(rubric: 16 pts)*
`.github/workflows/pipeline.yml` automates the end-to-end workflow with **one job per ML step**:

| Job | What it does | Hand-off |
|---|---|---|
| `register-dataset` | runs `data_register.py`; fails the run if the schema is wrong | — |
| `data-prep` | runs `prep.py` | uploads **`train-test-splits`** artifact |
| `model-training` | downloads the artifact, runs `train.py` (GridSearch + MLflow) | uploads `mlflow-tracking` + `model` artifacts, **commits the new model and pushes to `main`** |

**Triggers:** every push to `main` that touches the data, code or workflow, plus manual `workflow_dispatch`.
**Auto-push:** the training job commits `tourism_project/model/` back to `main` using the built-in `GITHUB_TOKEN`
(`permissions: contents: write`). The commit message carries `[skip ci]`, so the bot's own push does not re-trigger the pipeline.
Each new model commit makes Streamlit Community Cloud redeploy the app automatically — closing the CI/CD loop.

In [17]:
%%writefile .github/workflows/pipeline.yml
name: Tourism Package Prediction - MLOps Pipeline

on:
  push:
    branches: [main]
    paths:
      - "tourism_project/data/tourism.csv"
      - "tourism_project/*.py"
      - "tourism_project/requirements.txt"
      - ".github/workflows/pipeline.yml"
  workflow_dispatch:

permissions:
  contents: write          # allows the pipeline to push the trained model to main

env:
  PYTHON_VERSION: "3.12"

jobs:
  register-dataset:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: ${{ env.PYTHON_VERSION }}
          cache: pip
          cache-dependency-path: tourism_project/requirements.txt
      - run: pip install -r tourism_project/requirements.txt
      - name: Validate & register dataset
        run: python tourism_project/data_register.py

  data-prep:
    needs: register-dataset
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: ${{ env.PYTHON_VERSION }}
          cache: pip
          cache-dependency-path: tourism_project/requirements.txt
      - run: pip install -r tourism_project/requirements.txt
      - name: Clean data and create train/test splits
        run: python tourism_project/prep.py
      - name: Upload splits for the training job
        uses: actions/upload-artifact@v4
        with:
          name: train-test-splits
          path: tourism_project/data/processed/
          retention-days: 7

  model-training:
    needs: data-prep
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: actions/setup-python@v5
        with:
          python-version: ${{ env.PYTHON_VERSION }}
          cache: pip
          cache-dependency-path: tourism_project/requirements.txt
      - run: pip install -r tourism_project/requirements.txt
      - name: Download splits artifact
        uses: actions/download-artifact@v4
        with:
          name: train-test-splits
          path: splits/
      - name: Tune, track (MLflow) and save best model
        env:
          ARTIFACT_DIR: splits
          MLFLOW_TRACKING_URI: sqlite:///mlflow.db
        run: python tourism_project/train.py
      - name: Upload MLflow tracking store
        uses: actions/upload-artifact@v4
        with:
          name: mlflow-tracking
          path: |
            mlflow.db
            mlruns/
      - name: Upload trained model
        uses: actions/upload-artifact@v4
        with:
          name: model
          path: tourism_project/model/
      - name: Commit and push the best model to main
        run: |
          git config user.name  "github-actions[bot]"
          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
          git add tourism_project/model/
          if git diff --cached --quiet; then
            echo "Model unchanged - nothing to commit"
          else
            git commit -m "ci: update trained model (run ${{ github.run_number }}) [skip ci]"
            git pull --rebase origin main
            git push origin HEAD:main
          fi

Writing .github/workflows/pipeline.yml


In [18]:
%%writefile tourism_project/requirements.txt
pandas==3.0.5
numpy==2.5.3
scikit-learn==1.9.1
xgboost==3.4.1
joblib==1.6.0
mlflow==3.16.1

Writing tourism_project/requirements.txt


In [19]:
%%writefile .gitignore
__pycache__/
*.pyc
.ipynb_checkpoints/
mlflow.db
mlruns/
mlartifacts/
tourism_project/data/processed/
splits/

Writing .gitignore


In [20]:
%%writefile README.md
# Tourism Package Prediction — MLOps Pipeline

Predicts whether a "Visit with Us" customer will buy the **Wellness Tourism Package** before they are contacted.

| Step | File |
|---|---|
| Data registration | `tourism_project/data_register.py` |
| Data preparation  | `tourism_project/prep.py` |
| Training + MLflow | `tourism_project/train.py` |
| Trained model (committed by CI) | `tourism_project/model/best_model.joblib` |
| Streamlit app | `tourism_project/deployment/app.py` |
| CI/CD pipeline | `.github/workflows/pipeline.yml` |

Every push to `main` runs: register → prep (artifact) → train (artifact in, model committed back to `main`).
Streamlit Community Cloud redeploys the app on each new model commit.

**Live app:** https://<your-app>.streamlit.app

Run locally:
```bash
pip install -r tourism_project/requirements.txt
python tourism_project/data_register.py && python tourism_project/prep.py && python tourism_project/train.py
streamlit run tourism_project/deployment/app.py
```

Writing README.md


### 7.1 Validate the workflow file and dry-run every job in order

In [21]:
import yaml
wf = yaml.safe_load(open(".github/workflows/pipeline.yml"))
for name, job in wf["jobs"].items():
    print(f"{name:18s} needs={job.get('needs', '-'):18s} steps={[s.get('name') or s.get('uses') or s.get('run') for s in job['steps']]}")

register-dataset   needs=-                  steps=['actions/checkout@v4', 'actions/setup-python@v5', 'pip install -r tourism_project/requirements.txt', 'Validate & register dataset']
data-prep          needs=register-dataset   steps=['actions/checkout@v4', 'actions/setup-python@v5', 'pip install -r tourism_project/requirements.txt', 'Clean data and create train/test splits', 'Upload splits for the training job']
model-training     needs=data-prep          steps=['actions/checkout@v4', 'actions/setup-python@v5', 'pip install -r tourism_project/requirements.txt', 'Download splits artifact', 'Tune, track (MLflow) and save best model', 'Upload MLflow tracking store', 'Upload trained model', 'Commit and push the best model to main']


In [22]:
# Local dry-run of the three jobs exactly as CI chains them (artifact copied to splits/ like download-artifact)
!rm -rf splits mlflow.db && python tourism_project/data_register.py > /dev/null && echo "job 1 register-dataset: OK"
!python tourism_project/prep.py > /dev/null && cp -r tourism_project/data/processed splits && echo "job 2 data-prep: OK (artifact -> splits/)"
!ARTIFACT_DIR=splits python tourism_project/train.py | tail -2 && echo "job 3 model-training: OK"

job 1 register-dataset: OK


job 2 data-prep: OK (artifact -> splits/)


2026/10/05 04:19:46 INFO mlflow.store.db.utils: Creating initial MLflow database tables...
2026/10/05 04:19:46 INFO mlflow.store.db.utils: Updating database tables


2026/10/05 04:19:47 INFO mlflow.tracking.fluent: Experiment with name 'tourism-package-prediction' does not exist. Creating a new experiment.



Saved best model to /home/claude/repo/tourism_project/model/best_model.joblib
job 3 model-training: OK


### 7.2 Push everything to GitHub
Create an empty public repo `tourism-mlops` on GitHub, then run the cell below (Colab: store a personal access token
with `repo` + `workflow` scopes in the Secrets panel as `GITHUB_TOKEN`). The push triggers the pipeline automatically.

In [23]:
# Set to True after filling in your username and token to push from the notebook
PUSH = False
GH_USER, REPO = "<your-username>", "tourism-mlops"
if PUSH:
    from google.colab import userdata
    token = userdata.get("GITHUB_TOKEN")
    !git init -b main && git add . && git commit -m "Tourism package prediction MLOps pipeline"
    !git remote add origin https://{GH_USER}:{token}@github.com/{GH_USER}/{REPO}.git && git push -u origin main
else:
    print("Skipped push (PUSH=False). Repository contents ready to push:")
    !find . -path ./.git -prune -o -path ./splits -prune -o -path ./mlruns -prune -o -path ./tourism_project/data/processed -prune -o -name '*.db' -prune -o -name '.ipynb_checkpoints' -prune -o -type f -print | grep -v -E "ipynb|\.html|^\./tourism.csv$|__pycache__" | sort

Skipped push (PUSH=False). Repository contents ready to push:
./.github/workflows/pipeline.yml
./.gitignore
./README.md
./tourism_project/data/tourism.csv
./tourism_project/data_register.py
./tourism_project/deployment/app.py
./tourism_project/deployment/requirements.txt
./tourism_project/model/best_model.joblib
./tourism_project/model/metrics.json
./tourism_project/prep.py
./tourism_project/requirements.txt
./tourism_project/train.py


### 7.3 Deploy on Streamlit Community Cloud
1. Go to **share.streamlit.io** → *Create app* → *Deploy a public app from GitHub*.
2. Repository: `<your-username>/tourism-mlops`, Branch: `main`, Main file path: `tourism_project/deployment/app.py`.
3. *Advanced settings* → Python **3.12** (same as CI, so the pickled model loads with identical library versions).
4. Deploy. Every model commit pushed by the pipeline triggers an automatic redeploy.

## 8. Output Evaluation *(rubric: 4 pts)*

**GitHub repository:** `https://github.com/<your-username>/tourism-mlops`

**Folder structure** *(screenshot)*

```
tourism-mlops/
├── .github/workflows/pipeline.yml
├── README.md
├── .gitignore
└── tourism_project/
    ├── data/tourism.csv
    ├── data_register.py
    ├── prep.py
    ├── train.py
    ├── requirements.txt
    ├── model/
    │   ├── best_model.joblib      ← committed by GitHub Actions
    │   └── metrics.json
    └── deployment/
        ├── app.py
        └── requirements.txt
```

`![Folder structure](screenshots/github_folder_structure.png)`

**Executed workflow** *(screenshot of the Actions run showing the 3 green jobs)*

`![Workflow run](screenshots/github_actions_run.png)`

**Streamlit app:** `https://<your-app>.streamlit.app`

`![Streamlit app](screenshots/streamlit_app.png)`

## 9. Insights & Recommendations
### 9.1 Conversion rate by key segments (evidence for the insights below)

In [24]:
df = pd.read_csv("tourism_project/data/tourism.csv").replace({"Gender": {"Fe Male": "Female"}})
df["AgeBand"] = pd.cut(df["Age"], [17, 25, 30, 40, 50, 61])
for col in ["Passport", "MaritalStatus", "Designation", "AgeBand", "CityTier", "NumberOfFollowups", "PreferredPropertyStar"]:
    print(df.groupby(col, observed=True)["ProdTaken"].agg(conversion="mean", customers="size").round(3).T.to_string(), "\n")

Passport           0         1
conversion     0.124     0.358
customers   2909.000  1219.000 

MaritalStatus  Divorced   Married   Single  Unmarried
conversion        0.133     0.142    0.364      0.243
customers       789.000  1990.000  667.000    682.000 

Designation      AVP  Executive   Manager  Senior Manager       VP
conversion     0.076        0.3     0.116           0.161    0.087
customers    250.000     1615.0  1422.000         737.000  104.000 

AgeBand     (17, 25]  (25, 30]  (30, 40]  (40, 50]  (50, 61]
conversion     0.433     0.266     0.154     0.146     0.161
customers    314.000   703.000  1776.000   900.000   435.000 

CityTier           1        2         3
conversion     0.165    0.265     0.241
customers   2678.000  162.000  1288.000 

NumberOfFollowups      1.0      2.0       3.0       4.0      5.0      6.0
conversion           0.132    0.102     0.168     0.189    0.249    0.395
customers          144.000  206.000  1157.000  1811.000  686.000  124.000 

Preferr

### 9.2 Summary

**Model**
* The tuned XGBoost pipeline scores **test F1 ≈ 0.80, precision ≈ 0.83, recall ≈ 0.77, ROC-AUC ≈ 0.94** (accuracy ≈ 0.93 against a 0.81 "nobody buys" baseline). It ranks likely buyers far better than random calling.
* `scale_pos_weight` pushes the model toward catching buyers; a missed buyer costs a sale while an extra call costs a few minutes, so recall matters more than precision here.
* Train scores are near-perfect (F1 ≈ 0.99) against test F1 ≈ 0.80, so the model over-fits somewhat. The pipeline makes it cheap to retrain with stronger regularisation (shallower trees, `min_child_weight`, `reg_lambda`) as more data arrives.

**Business (from feature importance and the segment table above)**
* **Product pitched / designation** are the strongest drivers: Basic-product (Executive) customers convert at **30 %**, against 8–16 % for the other tiers.
* **Passport holders** convert at **36 %** against 12 % without one, roughly 3x. Prioritise them.
* **Single** (36 %) and **unmarried** (24 %) customers convert far more than married (14 %) or divorced (13 %) ones.
* **Younger customers** convert best: **43 %** for ages 18–25 and 27 % for 26–30, falling to about 15 % above 30.
* **Tier-2/3 cities** (24–27 %) beat Tier-1 (17 %). Customers preferring **5-star** properties convert at 27 % against 17 % for 3-star.
* **Persistence pays:** conversion climbs from about 10–13 % with 1–2 follow-ups to **25 % at 5 and 40 % at 6**, and longer pitches convert better too. The sales team should put more follow-ups into high-score leads rather than spreading effort evenly.
* Recommended use: score every new lead, call the top-scored segment first, and track campaign conversion monthly against the model's predictions.

**MLOps**
* Every data or code change automatically re-validates the data, rebuilds splits, retunes the model, logs all experiments and ships the new model to the live app — no manual step between a new dataset and an updated production model.
* Next steps: add a quality gate (fail the run if test F1 drops below the current model's), data-drift checks on incoming data, and a scheduled weekly retrain.